### ISSUE TITLE
Libmemcached cache\backend performance regression

### ISSUE THREAD

**Issue by patrick-zippenfenig** [phalcon/cphalcon#10343]
Hi!

tl;dr: please disable "statKey" (storing ALL keys) per default

I just spend 6 hours debugging a performance regression in memcache. One of three memcached servers updated keys at to up 130Mbit/s. In the last couple of days totals >5TB. Network bandwidth showed an odd pattern:
![unknown-3](https://cloud.githubusercontent.com/assets/11670456/7615862/abea6010-f9a0-11e4-9ec1-7ed83f21a8f6.jpeg)

I suspected the issue anywhere but in phalcon as it is a quite complex webapplication with a lot of objects in memcached. I started logging cache set/get's, changed cache keys to prevent possible uneven distribution, binary protocol, compression, implemented persistent memcached connections, analysed access-logs, etc.

After tcp_dump'ing memcache packets and analysing traffic flow I recognised "_PHCM" was updated very frequently. A quick look at [libmemcached.zep#L87](https://github.com/phalcon/cphalcon/blob/master/phalcon/cache/backend/libmemcached.zep#L87) reveals: "statKey" feature is enabled per default. Every updated key is stored as an entry in memcached to enable ````$cache->flush()```` capabilities. ````"statKey" => ""```` solved the issue, dropped load by half, freed a few GB of memory (memcache+php) and reduced network bandwidth to a regular level.
Phalcon queried "_PHCM" for each ````$cache->save()````, added the modified key and stored it again in "_PHCM". After some time time an internal limit was hit, crashed and it started all over again (see bandwidth graph).

I doubt that any application use ````flush()```` in production. I would like to request to disable this behaviour per default in all cache backends. An exception is already thrown in ````flush()```` if "statKey" is disabled.

Thanks!

**Comment by andresgutierrez** [comment]
+1 for disabling statKey by default

**Comment by kimthangatm** [comment]
Hi, I use Phalcon v2.0, Code `$cache->flush()` not working.

**Comment by dugwood** [comment]
@patrick-zippenfenig : I fixed it a while ago (https://github.com/phalcon/cphalcon/pull/3183) but nobody seemed to be ashamed of such an «issue»

@kimthangatm : flush() won't work, as it just empty all the known cached entries, but keeps the list. My suggestion: add a parameter in the prefix of Memcached parameters (Memcached::OPT_PREFIX_KEY), and update the prefix when you want all your keys to be disabled. That won't remove them, but after the timeout these could be removed automatically.

I must point out that removing the values from Memcached is a BAD idea: it has never been implemented this way (and there no way to list all Memcached keys too!).

**Comment by patrick-zippenfenig** [comment]
@dugwood I did not notice #3183. A token/timestamp would be a good solution for this feature, but still has negative performance impact. I use multiple revision-numbers as prefixes for my cache-keys which are hard-coded in my application. For rolling releases I manually increment certain revision-numbers to invalidate parts of my cached data.

**Comment by dugwood** [comment]
Yes, the idea was for @kimthangatm who's using flush() (this function should trigger Memcached::flush(), else it should be completely removed).

And I must confess that I have the same system for revision as you do, either for objects, dedicated cache, and so on. I use timestamp, so when it's necessary to update the cache, it set the revision to the current timestamp using time() or $_SERVER['REQUEST_TIME']. Using revision numbers is okay if you store the numbers in a safe place, else you may restart at 1 and end up on very old caches.

**Comment by sergeyklay** [comment]
Hello.
Confirm for 2.0.7.
Recently I began to use Libmemcached as backend on my production server. Before I used native Memcache. Now I have load average 40 on 8 core cpu (previously, it was 15).

**Comment by dugwood** [comment]
Did you try to disable statsKey with options["statsKey"] = "" (empty string)?

Memcached is better than Memcache in general, if you have a load increase that may come from another issue: not enough space in RAM, or a new parameter that goes wrong for you (increased timeout).
I don't think the issue is Memcached anyway, you should profile you application (try xhprof for instance), you'll know for sure.

**Comment by sergeyklay** [comment]
@dugwood Yes. I disavled last night statsKey. After I turned off statsky everything fell into place.

**Comment by sergeyklay** [comment]
After disabling statsKey

![selection_114](https://cloud.githubusercontent.com/assets/1256298/9780965/c3cbfcac-5797-11e5-9c5b-bccd6fc98402.png)

**Comment by dugwood** [comment]
@sergeyklay :-)

I've detected it when I ran a script with about 10,000 objects, all from cache, and the time for getting the object from cache was over the time from creating it from the database.

@andresgutierrez : did you disabled it as a default? Note that I've added a warning for anyone who's using «flush()» when statsKey is disable, so there's no harm doing this (I think that a lot of people will say that phalcon is faster on cache actions after this change :-)

**Comment by sergeyklay** [comment]
@dugwood 
> I've detected it when I ran a script with about 10,000 objects, all from cache, and the time for getting the object from cache was over the time from creating it from the database.

It's time to create Phalcon handlersocket wrapper which is usually faster than memcache at all :-)

**Comment by patrick-zippenfenig** [comment]
@dugwood On 2.1.x stats keys are now disabled by default (#10841). I've adjusted your warning accordingly.

I also use persistent memcached pool connections for cache and sessions. I'll port it to zephir and open a PR
````php
$memcache = new \Memcached('my-pool');

if (empty($memcache->getServerList())) {
  $memcache->addServers($options['servers']);
  $memcache->setOptions($options['client']);
}
````

**Comment by dugwood** [comment]
Thanks @patrick-zippenfenig, I should upgrade to 2.1 then :-)

For Memcached, here's my config (within the DI):
```php
return new Libmemcached(new Data(['lifetime' => 3600]), [
				'servers' => $servers,
				'client' => [
					/* 100 ms de timeout de connexion car le réseau OVH est super lent ces dernières semaines */
					Memcached::OPT_CONNECT_TIMEOUT => 100,
					/* Distribution indépendante du serveur (si HS) */
					Memcached::OPT_DISTRIBUTION => Memcached::DISTRIBUTION_CONSISTENT,
					/* On supprime les serveurs HS de la liste des serveurs quand ils sont marqués «DEAD» */
					Memcached::OPT_REMOVE_FAILED_SERVERS => true,
					/* Nombre de tentatives avant de marquer un serveur comme HS */
					Memcached::OPT_SERVER_FAILURE_LIMIT => 2,
					/* On autorise les serveurs HS à être testés à nouveau au bout de 1 seconde */
					Memcached::OPT_RETRY_TIMEOUT => 1
				],
				'statsKey' => '']);
```

Here's some explanations on my website: http://www.dugwood.com/895442.html

**Comment by sergeyklay** [comment]
One of my node (frontend) after disabling statsKey
![selection_009](https://cloud.githubusercontent.com/assets/1256298/9785858/2380e952-57be-11e5-99f0-3bb3e0b6fb10.png)

**Comment by sergeyklay** [comment]
After enabling `Libmemcached` with `statsKey`
![selection_010](https://cloud.githubusercontent.com/assets/1256298/9788744/80b34a14-57d4-11e5-88cc-7a4f6632cf48.png)

**Comment by patrick-zippenfenig** [comment]
Just noticed ````setOptions```` is called AFTER ````addServers```` in current phalcon [memcached implementation](https://github.com/phalcon/cphalcon/blob/2.1.x/phalcon/cache/backend/libmemcached.zep#L112) -> Some options are ignored....

Currently working on a PR with persistent libmemcached and sane default options

**Comment by tylercb** [comment]
Swapping out memcached for redis fixed the performance problems I experienced related to this.

![memcache vs redis 2](https://cloud.githubusercontent.com/assets/5401679/10974249/cefab92a-83a6-11e5-9497-a88dd35807ce.png)

**Comment by louiemiranda** [comment]
Reviewing comments and code later, but need a quick info would anyone know if this problem was resolved on 2.0.13?

**Comment by sergeyklay** [comment]
I'm not sure
